### PR TITLE
Defensively copy state vector in simulator

### PR DESCRIPTION
**Pull request by dabacon**
The state vector simulator version of #3109  One difference is that the code process the initial state already make a copy of the state.

### CODE REVIEW

File: cirq/sim/state_vector_simulator.py [hunk 1]
```diff
@@ -212,7 +212,7 @@ def state_vector(self):
                 |  6  |   1    |   1    |   0    |
                 |  7  |   1    |   1    |   1    |
         """
-        return self._final_simulator_state.state_vector
+        return self._final_simulator_state.state_vector.copy()
```
Comment (95-martin-orion) [inline cirq/sim/state_vector_simulator.py:215]: Should we also have `test_final_state_vector_is_not_last_object` for this object?

File: cirq/sim/sparse_simulator_test.py [hunk 2]
```diff
@@ -963,3 +964,46 @@ def test_separated_measurements():
...
+    q = cirq.LineQubit(0)
+    circuit = cirq.Circuit(InplaceGate()(q), InplaceGate()(q))
+
+    vectors = []
+    for step in sim.simulate_moment_steps(circuit):
+        vectors.append(step.state_vector(copy=True))
+    for x, y in itertools.combinations(vectors, 2):
+        assert not np.shares_memory(x, y)
+
+    # If the state vector is not copied, then applying second Hadamard
```
Comment (95-martin-orion) [inline cirq/sim/sparse_simulator_test.py:988]: `Hadamard` -> `InplaceGate`

Comment (dabacon) [reply cirq/sim/sparse_simulator_test.py:988]: Done

File: cirq/sim/state_vector_simulator.py (hunk 1 above)
Comment (dabacon) [reply cirq/sim/state_vector_simulator.py:215]: It is actually tested in sparse_vector_simulator_test (result.state_vector). But I've added a test to state_vector_simulator_test to test the copy here.

**Comment by dabacon** [comment]
PTAL

**Pull request merged**